### PR TITLE
✨(models) add marsha video events pydantic models

### PR DESCRIPTION
**Pull request by quitterie-lcs**
## Purpose

Video events from `marsha` will be tracked for `Potsie` dashboards. We implement their pydantic models to stream them in `ralph`.

## Proposal

- [x] `initialized`
- [x] `paused`
- [x] `seeked`
- [x] `completed`
- [x] `terminated`
- [x] `interacted`

(`played` already implemented in PR #96)

### CODE REVIEW

File: src/ralph/models/xapi/video/constants.py [hunk 1]
```diff
@@ -5,12 +5,47 @@
...
+)
+VIDEO_EXTENSION_FRAME_RATE = "https://w3id.org/xapi/video/extensions/frame-rate"
+VIDEO_EXTENSION_FULL_SCREEN = "https://w3id.org/xapi/video/extensions/full-screen"
+VIDEO_EXTENSION_QUALITY = "https://w3id.org/xapi/video/extensions/quality"
+VIDEO_EXTENSION_SCREEN_SIZE = "https://w3id.org/xapi/video/extensions/screen-size"
+VIDEO_EXTENSION_VIDEO_PLAYBACK_SIZE = (
+    "https://w3id.org/xapi/video/extensions/video-playback-size"
+)
+VIDEO_EXTENSION_SPEED = "https://w3id.org/xapi/video/extensions/speed"
+VIDEO_EXTENSION_TRACK = "https://w3id.org/xapi/video/extensions/track"
```
Comment (quitterie-lcs) [inline src/ralph/models/xapi/video/constants.py:30]: Not defined in xapi

Comment (SergioSim) [reply src/ralph/models/xapi/video/constants.py:30]: Do you search for this extension  https://liveaspankaj.gitbooks.io/xapi-video-profile/content/statement_data_model.html#26210-track ?

Comment (quitterie-lcs) [reply src/ralph/models/xapi/video/constants.py:30]: That's it, yes! Thanks! Should not have written it openly, it was notes in my WIP, sorry...

**Review by SergioSim — COMMENTED** [review]
This is great! Thank you for your contribution! :+1:  Have only some minor suggestions)

File: src/ralph/models/xapi/video/statements.py [hunk 2]
```diff
@@ -1,25 +1,150 @@
...
+
+    Attributes:
+        object (dict): See VideoObjectField.
+    """
+
+    object: VideoObjectField
+
+
+class VideoInitialized(BaseVideoStatement):
+    """Represents a video initialized xAPI statement.
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/statements.py:44]: Can you explain a bit more what every event correspond to? E.g. _John launched the video player_

**Review by jmaupetit — COMMENTED** [review]
Nice work!

I tried to validate real world statements using your model definitions for video statements and I found that we should add minor improvements to this work/ralph:

1. We should define the `__selector__` attribute for every xAPI model (video + navigation),
2. The `ralph.models.xapi` module should define every active model (for reference see https://github.com/openfun/ralph/blob/master/src/ralph/models/edx/__init__.py),
3. The `validate` command should add `xapi` to supported formats, and,
4. The `ModelSelector` should raise an exception if no rules have been found for a format*.

\* This point can be handled in a new PR (ping @SergioSim :wink:)

File: src/ralph/models/xapi/video/statements.py [hunk 3]
```diff
@@ -1,25 +1,150 @@
...
+
+    WARNING: This model is based on the video initialized xAPI event generated by `marsha`.
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/xapi/video/statements.py:46]: I wanted to specifiy that the extensions/fields declared are specific to `marsha`.

File: src/ralph/models/xapi/video/fields/results.py [hunk 4]
```diff
@@ -1,28 +1,151 @@
...
 
     time: Optional[float] = Field(alias=VIDEO_EXTENSION_TIME)
 
 
+class VideoPausedResultExtensionsField(VideoActionResultExtensionsField):
+    """Represents the result.extensions field for video `played` xAPI statement.
+
+    Attributes:
+        playedSegments (str): Consists of parts of the video the actor watched during current
+            registration in chronological order.
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/xapi/video/fields/results.py:34]: > Nice work!
> 
> I tried to validate real world statements using your model definitions for video statements and I found that we should add minor improvements to this work/ralph:
> 
> 1. We should define the `__selector__` attribute for every xAPI model (video + navigation),

It's a miss on my part. Thanks! 

> 2. The `ralph.models.xapi` module should define every active model (for reference see https://github.com/openfun/ralph/blob/master/src/ralph/models/edx/__init__.py),

:+1: 

> 3. The `validate` command should add `xapi` to supported formats, and,

Indeed. Thanks! 

> 4. The `ModelSelector` should raise an exception if no rules have been found for a format*.
> 
> * This point can be handled in a new PR (ping @SergioSim )

**Comment by quitterie-lcs** [comment]
> Nice work!
> 
> I tried to validate real world statements using your model definitions for video statements and I found that we should add minor improvements to this work/ralph:
> 
> 1. We should define the `__selector__` attribute for every xAPI model (video + navigation),

It's a miss on my part. Thanks! 

> 2. The `ralph.models.xapi` module should define every active model (for reference see https://github.com/openfun/ralph/blob/master/src/ralph/models/edx/__init__.py),

:+1: 

> 3. The `validate` command should add `xapi` to supported formats, and,

Indeed. Thanks! 

> 4. The `ModelSelector` should raise an exception if no rules have been found for a format*.
> 
> * This point can be handled in a new PR (ping @SergioSim )

File: src/ralph/models/xapi/__init__.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+from .video.statements import (
+    VideoCompleted,
+    VideoInitialized,
+    VideoInteracted,
+    VideoPaused,
+    VideoPlayed,
+    VideoSeeked,
+    VideoTerminated,
+)
```
Comment (jmaupetit) [inline src/ralph/models/xapi/__init__.py:13]: You can also add navigation events. WDYT?

File: src/ralph/models/xapi/video/statements.py [hunk 6]
```diff
@@ -1,25 +1,179 @@
...
+class VideoInitialized(BaseVideoStatement):
+    """Represents a video initialized xAPI statement.
+
+    Example: A video has been fully initialized.
+
+    Attributes:
+        verb (dict): See VideoInitializedVerbField.
+        context (dict): See VideoInitializedContextField.
+    """
+
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/statements.py:53]: The selector is missing here.

File: src/ralph/models/xapi/__init__.py (hunk 5 above)
Comment (jmaupetit) [reply src/ralph/models/xapi/__init__.py:13]: And selector for those events (maybe in a separated commit).

File: src/ralph/models/xapi/video/statements.py [hunk 7]
```diff
@@ -1,25 +1,179 @@
...
+    verb: VideoInitializedVerbField
+    context: VideoInitializedContextField
+
+
+class VideoPlayed(BaseVideoStatement):
+    """Represents a video played xAPI statement.
+
+    Example: John played the video or clicked the play button.
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/statements.py:61]:
```suggestion
    Example: John played the video or clicked the play button.
    
```

File: src/ralph/models/xapi/video/statements.py [hunk 8]
```diff
@@ -1,25 +1,179 @@
...
+
     verb: VideoPlayedVerbField
     result: VideoPlayedResultField
     context: VideoPlayedContextField
+
+
+class VideoPaused(BaseVideoStatement):
+    """Represents a video paused xAPI statement.
+
+    Example: John paused the video or clicked the pause button.
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/statements.py:81]:
```suggestion
    Example: John paused the video or clicked the pause button.
    
```

File: src/ralph/models/xapi/video/statements.py [hunk 9]
```diff
@@ -1,25 +1,179 @@
...
+
+    verb: VideoSeekedVerbField
+    result: VideoSeekedResultField
+    context: VideoSeekedContextField
+
+
+class VideoCompleted(BaseVideoStatement):
+    """Represents a video completed xAPI statement.
+
+    Example: John completed a video by watching all parts of the video at least once
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/statements.py:122]:
```suggestion
    Example: John completed a video by watching major parts of the video at least once
    
```

File: src/ralph/models/xapi/video/statements.py [hunk 10]
```diff
@@ -1,25 +1,179 @@
...
+
+    verb: VideoCompletedVerbField
+    result: VideoCompletedResultField
+    context: VideoCompletedContextField
+
+
+class VideoTerminated(BaseVideoStatement):
+    """Represents a video terminated xAPI statement.
+
+    Example: John ended a video.
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/statements.py:142]:
```suggestion
    Example: John ended a video (quit the player).
```

**Review by jmaupetit — COMMENTED** [review]
Almost done. Keep it up :muscle:

File: src/ralph/models/xapi/base.py [hunk 11]
```diff
@@ -15,11 +14,11 @@ class BaseXapiModel(BaseModelWithConfig):
     `context`, `result`, `stored`, `authority`, `version` and `attachments` fields.
 
     Attributes:
-        id (UUID5): Consists of a generated UUID5 from the source event string.
+        id (UUID4): Consists of a generated UUID4 from the source event string.
```
Comment (jmaupetit) [inline src/ralph/models/xapi/base.py:17]:
```suggestion
        id (UUID): Consists of a generated UUID string from the source event string.
```

File: src/ralph/models/xapi/video/fields/results.py [hunk 12]
```diff
@@ -1,28 +1,164 @@
...
+)
 
 
-class VideoPlayedResultExtensionsField(BaseModelWithConfig):
+class VideoActionResultExtensionsField(BaseModelWithConfig):
     """Represents the result.extensions field for video `played` xAPI statement.
 
     Attributes:
-        time (float): Consists of the spent time on the video when the event was emitted.
+        time (float): Consists of the time into the video.
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/fields/results.py:23]:
```suggestion
        time (float): Consists of the video time code when the event was emitted.
```

File: src/ralph/models/xapi/video/fields/results.py [hunk 13]
```diff
@@ -1,28 +1,164 @@
...
 
 
+class VideoPausedResultExtensionsField(VideoActionResultExtensionsField):
+    """Represents the result.extensions field for video `played` xAPI statement.
+
+    Attributes:
+        playedSegments (str): Consists of parts of the video the actor watched during current
+            registration in chronological order (for example,
+            "0[.]5[,]12[.]22[,]15[.]55[,]55[.]99.33[,]99.33").
+        progress (float): Consists of the percentage of media consumed by the actor.
```
Comment (jmaupetit) [inline src/ralph/models/xapi/video/fields/results.py:36]:
```suggestion
        progress (float): Consists of the ratio of media consumed by the actor.
```

**Review by jmaupetit — APPROVED** [review]
LGTM

**Pull request merged**